Linux #2

rhowe opened the pull request:
I've taken https://github.com/mastahyeti/certstore/pull/11 and (hopefully) addressed all of the comments there.

Please bear in mind my go experience is minimal, so review ruthlessly.

rhowe committed:
Add code from https://github.com/mastahyeti/certstore/pull/11

rhowe committed:
Run travis builds on Linux as well as OS X

rhowe committed:
Use pkg-config for build flags

rhowe committed:
Make use of fmt.Errorf()

rhowe committed:
Remove extraneous variable

rhowe committed:
Simplify code to retrieve user's home directory

rhowe committed:
Check that the NSS database path exists

rhowe committed:
Remove unnecessary casts

rhowe commented:
Also, the travis config for Linux isn't quite right

rhowe committed:
Implement crypto.Signer directly on nssIdentity

This means we can remove the nssPrivateKey type and duplicate certificate
parsing code.

rhowe committed:
Remove go wrappers for C macros

rhowe commented:
I had a go at redoing the building of the certificate chain but my lack of NSS knowledge is failing me. It's certainly not passing tests...

```
diff --git a/certstore_linux.go b/certstore_linux.go
index 0bf1d1d..d5bfa83 100644
--- a/certstore_linux.go
+++ b/certstore_linux.go
@@ -110,44 +110,23 @@ func (i *nssIdentity) Certificate() (*x509.Certificate, error) {
 }
 
 func (i *nssIdentity) CertificateChain() ([]*x509.Certificate, error) {
-	var der = i.node.cert.derCert
-	var bytes = C.GoBytes(unsafe.Pointer(der.data), C.int(der.len))
-	cert, err := x509.ParseCertificate(bytes)
-	if err != nil {
-		return nil, err
-	}
-	var certs = C.PK11_ListCerts(C.PK11CertListAll, nil)
-	if certs == nil {
-		return nil, fmt.Errorf("Error %d, cannot list certificates...\n", int(C.PR_GetError()))
+	var certs = C.CERT_GetCertChainFromCert(i.node.cert, C.PR_Now(), C.certUsageAnyCA)
+
+	var errno = int(C.PR_GetError())
+	if (errno != 0) {
+		return nil, fmt.Errorf("Error %d building certificate chain", errno)
 	}
-	var list *C.CERTCertList
-	var node *C.CERTCertListNode
-	list = certs
-	var identities = make([]Identity, 0)
+
 	var certificates = make([]*x509.Certificate, 0)
-	for node = C.CertListHead(list); C.CertListEnd(node, list) != 0; node = C.CertListNext(node) {
-		identities = append(identities, newNssIdentity(node))
-	}
-	certificates = append(certificates, cert)
-	found := true
-	for found != false {
-		found = false
-		cert = certificates[len(certificates)-1]
-		issuer := cert.Issuer.String()
-		for i := 0; i < len(identities); i++ {
-			cert, err = identities[i].Certificate()
-			if cert == nil {
-				return nil, fmt.Errorf("Error %d, cannot fetch certificate...\n", int(C.PR_GetError()))
-			}
-			subject := cert.Subject.String()
-			root := certificates[len(certificates)-1].Subject.String()
-			if subject == issuer && subject != root {
-				certificates = append(certificates, cert)
-				found = true
-				break
-			}
+	var node *C.CERTCertListNode
+	for node = C.CertListHead(certs); C.CertListEnd(node, certs) != 0; node = C.CertListNext(node) {
+		var cert, err = newNssIdentity(node).Certificate()
+		if err != nil {
+			return nil, errors.New("Unable to parse certificate when building chain")
 		}
+		certificates = append(certificates, cert)
 	}
+
 	return certificates, nil
 }
 ```

rhowe committed:
Make nssIdentity a type alias

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,336 @@
+type nssStore int
+
+func (nssStore) Identities() ([]Identity, error) {
+	var identities = make([]Identity, 0)
```
If you're defining multiple variables at once, it's better to use this syntax

```go
var (
  a = 1
  b = 2
)
```

Alternatively, if you're defining a single variable and setting it's value, it's preferable to use this syntax

```go
a := 1
```

rhowe replied:
OK, I think I've sorted these out

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,336 @@
+
+func (nssStore) Identities() ([]Identity, error) {
+	var identities = make([]Identity, 0)
+	//fmt.Printf("Listing certificates:\n")
```
I'd rather not merge this with commented out code.

rhowe replied:
Agreed. I've removed the commented-out debug code. I've left in some of the hash algorithm commented out code because I haven't looked at what it was trying to do yet.

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,336 @@
+
+type nssStore int
+
+func (nssStore) Identities() ([]Identity, error) {
```
All public methods need a documentation string. You can mostly just copy/paste from the platform implementations.

rhowe replied:
OK, I think I've added comments which make at least some vague amount of sense to me :)

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,336 @@
+	//fmt.Printf("Listing certificates:\n")
+	var certs = C.PK11_ListCerts(C.PK11CertListAll, nil)
+	if certs == nil {
+		C.NSS_Shutdown()
```
Do we really want to shutdown NSS on every error? I feel like this should maybe be done in `func (nssIdentity) Close()` instead.

rhowe replied:
Not sure, but see this comment on the original PR: https://github.com/mastahyeti/certstore/pull/11#discussion_r245041591

btoews replied:
Interesting. I think if we use the init-context function, then we can shutdown just the context we initialized without impacting other parts of the app using NSS.

rhowe replied:
Done

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,336 @@
+	return identities, nil
+}
+
+type nssIdentity C.CERTCertListNode
```
I feel like we should be using the `CERTCertificate` itself rather than the list entry for this type. It looks like we'll need to copy the cert with `CERT_DupCertificate` and then call `CERT_DestroyCertList` to free the list in `Identities`.

rhowe replied:
Yes, this makes sense. I'll have a crack at that

rhowe replied:
Done

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,336 @@
+	return cert, nil
+}
+
+func (i *nssIdentity) CertificateChain() ([]*x509.Certificate, error) {
```
I'm not reviewing this method, since it will hopefully be replaced with a call to `NSSCertificate_BuildChain`.

rhowe replied:
> I'm not reviewing this method, since it will hopefully be replaced with a call to `NSSCertificate_BuildChain`.

I used `CERT_GetCertChainFromCert` instead - it was simpler to use and seems to do the right thing...

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,336 @@
+}
+
+func (i *nssIdentity) Delete() error {
+	C.PK11_DeleteTokenCertAndKey(i.cert, nil)
```
This returns a `SECStatus` that we should probably check and turn into an error if unsuccessful.

rhowe replied:
True, although it does actually only ever return `SECSuccess` :)

rhowe replied:
Done

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,336 @@
+}
+
+func (i *nssIdentity) Public() (crypto.PublicKey) {
+	var cert, _ = i.Certificate()
```
There's probably a cheaper way to do this, but I'm fine with keeping the code here simple for now.

rhowe replied:
We could ask NSS to do it for us, but I think I prefer having this fully implemented in go

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,336 @@
+func (nssStore) Identities() ([]Identity, error) {
+	var identities = make([]Identity, 0)
+	//fmt.Printf("Listing certificates:\n")
+	var certs = C.PK11_ListCerts(C.PK11CertListAll, nil)
```
It seems like we should maybe be passing `PK11CertListUser` instead of `PK11CertListAll` to just get certificates that the private key is available for.

rhowe replied:
Yes, this makes sense - when I tested earlier and got all the system CA certificates I was a bit surprised :)

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,336 @@
+		return nil, errors.New("cannot find private key")
+	}
+	if key.keyType == C.rsaKey {
+		var pkcs1Prefix = map[crypto.Hash][]byte{
```
It feels a bit weird to be hardcoding ASN.1 encoded structures and concatenating it with the digest. I feel like this could use some refactoring.

rhowe replied:
I expect NSS can do all this for us.. just need to find out how

rhowe replied:
I think `Sign` should look something like this, but it's not producing quite the right results for me
```
// Sign implements the crypto.Signer interface.
func (i *nssIdentity) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
   hash := opts.HashFunc()
   if len(digest) != hash.Size() {
      return nil, errors.New("bad digest for hash")
   }
   var (
      cert    = C.CERTCertificate(*i)
      key     = C.PK11_FindKeyByAnyCert(&cert, nil)
      secitem C.SECItem
   )
   if key == nil {
      return nil, errors.New("cannot find private key")
   }
   goToNssAlg := map[crypto.Hash]C.SECOidTag{
      crypto.SHA1:   C.SEC_OID_SHA1,
      crypto.SHA256: C.SEC_OID_SHA256,
      crypto.SHA384: C.SEC_OID_SHA384,
      crypto.SHA512: C.SEC_OID_SHA512,
   }
   algID := C.SEC_GetSignatureAlgorithmOidTag(key.keyType, goToNssAlg[opts.HashFunc()])
   if algID == C.SEC_OID_UNKNOWN {
      return nil, errors.New("algorithm unknown")
   }
   rv    := C.SEC_SignData(&secitem, (*C.uchar)(unsafe.Pointer(&digest[0])), C.int(len(digest)), key, algID)
   if rv != C.SECSuccess {
      return nil, errors.New("unable to sign")
   }
   sig := C.GoBytes(unsafe.Pointer(secitem.data), C.int(secitem.len))
   return sig, nil
}
```

rhowe replied:
Or maybe
```
// Sign implements the crypto.Signer interface.
func (i *nssIdentity) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
   hash := opts.HashFunc()
   if len(digest) != hash.Size() {
      return nil, errors.New("bad digest for hash")
   }
   var (
      cert    = C.CERTCertificate(*i)
      key     = C.PK11_FindKeyByAnyCert(&cert, nil)
   )
   if key == nil {
      return nil, errors.New("cannot find private key")
   }
   goToNssAlg := map[crypto.Hash]C.SECOidTag{
      crypto.SHA1:   C.SEC_OID_SHA1,
      crypto.SHA256: C.SEC_OID_SHA256,
      crypto.SHA384: C.SEC_OID_SHA384,
      crypto.SHA512: C.SEC_OID_SHA512,
   }
   algID := C.SEC_GetSignatureAlgorithmOidTag(key.keyType, goToNssAlg[opts.HashFunc()])
   if algID == C.SEC_OID_UNKNOWN {
      return nil, errors.New("algorithm unknown")
   }
   ctx := C.SGN_NewContext(algID, key)
   if ctx == nil {
      return nil, errors.New("unable to initialise signing context")
   }
   rv := C.SGN_Begin(ctx)
   if rv != C.SECSuccess {
      return nil, errors.New("unable to initialise signing")
   }
   rv = C.SGN_Update(ctx, (*C.uchar)(unsafe.Pointer(&digest[0])), C.uint(len(digest)))
   if rv != C.SECSuccess {
      return nil, errors.New("unable to sign data")
   }
   var secitem C.SECItem
   rv = C.SGN_End(ctx, &secitem)
   if rv != C.SECSuccess {
      return nil, errors.New("unable to finalise signing")
   }
   sig := C.GoBytes(unsafe.Pointer(secitem.data), C.int(secitem.len))
   return sig, nil
}
``` but that also fails the tests

btoews replied:
My guess is that many of the "sign" functions are doing the digesting themselves. I'll do some reading to see if I can figure out the correct invocation here.

btoews replied:
I spent some time playing with this. I think we want to use `SGN_Digest` for signing. Here's the C program I was using for testing this

```C
#include <unistd.h>
#include <stdio.h>
#include <nss.h>
#include <secport.h>
#include <prerror.h>
#include <certt.h>
#include <cert.h>
#include <keyhi.h>
#include <cryptohi.h>
#include <pk11pub.h>

// Create database:
//   mkdir ~/.netscape && certutil -N

// Generate cert:
//   certutil -S -n test1 -x -7 mastahyeti@gmail.com -krsa -tu -s CN=Ben

// Compile:
//   gcc `pkg-config --cflags --libs nss` -o main ./main.c

int main(int argc, char** argv) {
  NSSInitContext *ctxt;
  CERTCertList *list;
  CERTCertificate *cert = NULL;
  SECKEYPrivateKey *key = NULL;
  SECStatus rv;
  SECItem it, digest;

  ctxt = NSS_InitContext("/Users/mastahyeti/.netscape", "", "", SECMOD_DB, NULL, 0);
  if (ctxt == NULL) {
    fprintf(stderr, "NSS_InitContext -> %s\n", PORT_ErrorToString(PORT_GetError()));
    return 1;
  }

  list = PK11_ListCerts(PK11CertListUser, NULL);
  if (list) {
    if (!CERT_LIST_EMPTY(list)) {
      cert = CERT_DupCertificate(CERT_LIST_HEAD(list)->cert);
    }
    CERT_DestroyCertList(list);
  }

  if (cert == NULL) {
    fprintf(stderr, "No cert found\n");
    goto done;
  }

  key = PK11_FindKeyByAnyCert(cert, NULL);
  if (key == NULL) {
    fprintf(stderr, "No private key found\n");
    goto done;
  }

  // Ruby - Digest::SHA256.hexdigest("hello, world!").chars.each_slice(2).map { |q| "0x#{q.join}" }.join(", ")
  unsigned char buf[] = { 0x68, 0xe6, 0x56, 0xb2, 0x51, 0xe6, 0x7e, 0x83, 0x58, 0xbe, 0xf8, 0x48, 0x3a, 0xb0, 0xd5, 0x1c, 0x66, 0x19, 0xf3, 0xe7, 0xa1, 0xa9, 0xf0, 0xe7, 0x58, 0x38, 0xd4, 0x1f, 0xf3, 0x68, 0xf7, 0x28 };
  digest.data = buf;
  digest.len = 32;
  digest.type = siBuffer;

  rv = SGN_Digest(key, SEC_OID_SHA256, &it, &digest);
  if (rv != SECSuccess) {
   fprintf(stderr, "Error signing: %s\n", PORT_ErrorToString(PORT_GetError()));
   goto done;
  }

  write(1, it.data, it.len);

done:
  SECKEY_DestroyPrivateKey(key);
  CERT_DestroyCertificate(cert);

  rv = NSS_ShutdownContext(ctxt);
  if (rv != SECSuccess) {
    fprintf(stderr, "NSS_ShutdownContext -> %d\n", rv);
  }
  return 0;
}
```

btoews replied:
Something different might need to be done for RSA PSS signatures. I haven't looked at that part yet.

rhowe replied:
Also EC signatures look like they're a bit different too.. 
I don't know if the SGN_* functions will use hardware tokens or not - need to have a play once it's all working

rhowe replied:
Great - I've managed to have some success! `SGN_Digest` is now passing tests.

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,336 @@
+			mechanism = C.CKM_SHA512_RSA_PKCS
+		*/
+		default:
+			mechanism = C.CKM_RSA_PKCS
```
I'm rusty enough on PKCS11 that I don't understand what the correct behavior is here. Does `CKM_SHA1_RSA_PKCS` imply that the input isn't already digested whereas `CKM_RSA_PKCS` implies that it is already digested?

rhowe replied:
This code is now gone

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,336 @@
+		return nil, errors.New("could not sign")
+	}
+	var sig = C.GoBytes(unsafe.Pointer(sd.data), C.int(sd.len))
+	if key.keyType == C.ecKey {
```
It feels like this method should be broken down into a few different methods that handle the special casing for different algorithms.

rhowe replied:
I think it's simple enough now that we can probably avoid it?

btoews replied:
Agreed!

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,336 @@
+	nssdbUrlC := C.CString(fmt.Sprintf("sql:/%s/", nssdb))
+	defer C.free(unsafe.Pointer(nssdbUrlC))
+	//fmt.Printf("Opening: %s\n", nssdbUrl)
+	ok := C.NSS_InitReadWrite(nssdbUrlC)
```
I wonder if we should be using `NSS_InitContext` instead to avoid messing with other libraries that might be using NSS.

rhowe replied:
Yes, I think we should... https://wiki.mozilla.org/NSS_Library_Init although more so that we can avoid calling `NSS_Shutdown`

rhowe replied:
OK, I've switched the code to use `NSS_InitContext`. It's maybe not optimal, but it passes tests :woman_shrugging:

btoews reviewed (commented):
I'm learning about NSS as I review, so I apologize for any misinformed comments/questions.

mastahyeti commented:
Thanks for picking this up, by the way :smile:

mastahyeti commented:
I added support for specifying `KeyUsage` with fakeca in https://github.com/mastahyeti/fakeca/pull/1 and pulled it into this repo in https://github.com/mastahyeti/certstore/pull/14. If you rebase onto master you should be able to test the chain building with KeyUsage set.

rhowe commented:
> Thanks for picking this up, by the way smile

No problem - thanks for all the comments. I'll get to them when I can :)

rhowe commented:
> I added support for specifying `KeyUsage` with fakeca in [mastahyeti/fakeca#1](https://github.com/mastahyeti/fakeca/pull/1) and pulled it into this repo in #14. If you rebase onto master you should be able to test the chain building with KeyUsage set.

hmm, looks like it isn't enough :(
I'll take a closer look at what the NSS code's actually doing

rhowe commented:
> I'm learning about NSS as I review, so I apologize for any misinformed comments/questions.

You and me both

rhowe committed:
Remove commented-out debugging code

rhowe committed:
Improve variable declaration style

rhowe committed:
Add comments for public methods

rhowe committed:
Only return certificates for which we have a private key

philandstuff committed:
style fixes

Errors should not start with a capital nor end with punctuation or
whitespace
https://github.com/golang/go/wiki/CodeReviewComments#error-strings

rhowe committed:
go fmt

rhowe committed:
Make nssIdentity a wrapper around a CERTCertificate

rhowe committed:
Handle potential errors from C.PK11_DeleteTokenCertAndKey

rhowe committed:
Use NSS to build certificate chain

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,343 @@
+// Identities implements the Store interface.
+func (nssStore) Identities() ([]Identity, error) {
+	var (
+		identities = make([]Identity, 0)
```
The `0` here isn't necessary. Just doing `make([]Identity)` will give you a slice with length and capacity of 0.

rhowe replied:
Is there more to it?
```
./certstore_linux.go:95:20: missing len argument to make([]Identity)
./certstore_linux.go:144:22: missing len argument to make([]*x509.Certificate)
```

btoews replied:
Don't mind me. I'm not sure what I was thinking there 🤦‍♂

btoews replied:
I guess the reason I was thinking this is that you don't usually see a second argument of `0`, though there's nothing wrong with it. When initializing an empty slice, I'll usually do `identities := []Identity{}` (functionally identical to what you're doing) or `var identities []Identity` (nil value, which works fine with `len()` and `append()`). Either way, what you're doing here is fine.

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,343 @@
+func (i *nssIdentity) CertificateChain() ([]*x509.Certificate, error) {
+	var (
+		cert  = C.CERTCertificate(*i)
+		certs = C.CERT_GetCertChainFromCert(&cert, C.PR_Now(), C.certUsageAnyCA)
```
We need to call `CERT_DestroyCertList` with this list when we're done with it.

btoews reviewed (commented):
I'm wondering if we should make NSS an option on all platforms, since it's cross platform already. Maybe we could define an `NSSOpen()` method on all platforms and then have `openStore()` call `NSSOpen()` on Linux (make NSS the default on Linux).

rhowe commented:
> I'm wondering if we should make NSS an option on all platforms, since it's cross platform already. Maybe we could define an `NSSOpen()` method on all platforms and then have `openStore()` call `NSSOpen()` on Linux (make NSS the default on Linux).

How about we get it working on one platform first? :joy: 

I'm less annoyed about my inability to get NSS to do what I want than I am about failing to get the travis build to install libnss-dev on linux prior to building.

rhowe committed:
Delegate signing more to NSS

Make use of SGN_Digest to perform our signing operations

rhowe committed:
Decrement the reference count on certs when we're done

We should be cleaning up after ourselves

rhowe committed:
Use NSS_InitContext/ShutdownContext

This makes us a nicer citizen

rhowe committed:
Remove commented-out debugging code

rhowe committed:
Support older NSS without PKCS#12 support for AES

rhowe commented:
> I'm less annoyed about my inability to get NSS to do what I want than I am about failing to get the travis build to install libnss-dev on linux prior to building.

Alright, Travis builds are working, but they do rely on an NSS database having been precreated in `~/.pki/nssdb`

rhowe committed:
Free password data after use in Import()

Fix this memory leak

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,318 @@
+}
+
+// nssStore is a bogus type. We have to explicitly open/close the store on
+// windows, so we provide those methods here too.
```
This comment isn't super accurate now that we're using `NSSInitContext`

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,318 @@
+		node       *C.CERTCertListNode
+	)
+	if certs == nil {
+		store.Close()
```
Is this a "fatal" error? It doesn't seem like closing the store should be necessary here.

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,318 @@
+	)
+	for node = C.CertListHead(certs); C.CertListEnd(node, certs) == 0; node = C.CertListNext(node) {
+		var (
+			identity  = nssIdentity(*C.CERT_DupCertificate(node.cert))
```
We're leaking these `CERTCertificate`s here. If we just remove the `CERT_DupCertificate` call, then these will be freed when `CERT_DestroyCertList` is called.

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,318 @@
+	return nil
+}
+
+// Close implements the Store interface.
```
This implements the _`Identity `_ interface

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,318 @@
+	defer C.SECITEM_FreeItem(resultC, 1)
+
+	digestC.len = C.uint(len(digest))
+	C.memcpy(unsafe.Pointer(digestC.data), unsafe.Pointer(&digest[0]), C.size_t(len(digest)))
```
Do you think it would be bad to skip the `SECITEM_AllocItem` call for `digest` and just do

```go
digestC := new(C.SECITEM)
digestC.len = C.uint(len(digest))
digestC.data = C.CBytes(digest)
```

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,318 @@
+	digestC.len = C.uint(len(digest))
+	C.memcpy(unsafe.Pointer(digestC.data), unsafe.Pointer(&digest[0]), C.size_t(len(digest)))
+
+	if C.SGN_Digest(key, goToNssAlg[hash], resultC, digestC) != C.SECSuccess {
```
The `goToNssAlg[hash]` lookup will panic on unsupported hash algorithms. Instead, let's do

```
nssAlg, ok := goToNssAlg[hash]
if !ok {
  return ni, errors.New("bad hash algo")
}

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,318 @@
+		return nil, fmt.Errorf("error signing: %s\n", C.GoString(C.GetErrorString()))
+	}
+
+	if key.keyType == C.ecKey {
```
We have to do this same conversion for Windows [here](https://github.com/mastahyeti/certstore/blob/bad86412360c8210ee584cdeb70e4d929ff48f40/certstore_windows.go#L419-L438). It might be nicer to extract that logic into a shared helper method and use it here as well to save another cgo call.

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,318 @@
+		return errors.New("SECITEM_AllocItem failed")
+	}
+	defer C.SECITEM_FreeItem(pass, 1)
+	C.memcpy(unsafe.Pointer(pass.data), unsafe.Pointer(&unicodePassword[0]), C.size_t(len(unicodePassword)))
```
Same here, for calling `C.CBytes` instead of using memcpy. I'm also wondering if a helper method would be nice

```go
func secitemFromBytes(b []byte) (secitem *C.SECITEM, free func()) {
  secitem = new(C.SECITEM)
  secitem.len = C.uint(len(b))
  secitem.data = C.CBytes(b)

  free = func() {
    C.SECITEM_FreeItem(secitem)
  }

  return
}
```

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,318 @@
+		p12     = C.SEC_PKCS12DecoderStart(pass, nil, nil, nil, nil, nil, nil, nil)
+		decoded = C.SEC_PKCS12DecoderUpdate(p12, (*C.uchar)(unsafe.Pointer(&data[0])), C.size_t(len(data)))
+	)
+	if decoded != 0 {
```
Let's check against `C.SECSuccess` for consistency.

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,318 @@
+	defer C.SECITEM_FreeItem(pass, 1)
+	C.memcpy(unsafe.Pointer(pass.data), unsafe.Pointer(&unicodePassword[0]), C.size_t(len(unicodePassword)))
+	var (
+		p12     = C.SEC_PKCS12DecoderStart(pass, nil, nil, nil, nil, nil, nil, nil)
```
It would be good to check the return value here.

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,318 @@
+		return nil, fmt.Errorf("NSS database not found at %s", nssdb)
+	}
+
+	ctx := C.NSS_InitContext(C.CString(nssdb), C.CString(""), C.CString(""), C.CString(""), nil, C.uint(0))
```
I think we're going to be leaking all these strings. Let's assign them to variables and free them in defer blocks.

btoews replied:
Thanks for switching to `NSS_InitContext `. This is feels _much_ nicer than using the global init/shutdown.

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,318 @@
+		return nil, err
+	}
+
+	nssdb := path.Join(homeDir, ".pki", "nssdb")
```
All the docs I'm seeing on macos say that the default directory is `~/.netscape`. It would be good to make this at least somewhat configurable. I'll open a separate PR to allow `openStore` to receive configuration arguments.

rhowe replied:
I'm not sure if it's this simple unfortunately - I don't have a `~/.netscape`, for example

See this:

```
$ find  ~/.* -type f -name cert9.db 2>/dev/null | sed -e 's#/home/[^/]\+#~#'
~/.icedove/<randomstring>.default/cert9.db
~/.mozilla/firefox/<randomstring>.default/cert9.db
~/snap/spotify/<random? number>/.pki/nssdb/cert9.db
```

I think applications are supposed to define their own directory for their user profile and dump it in there, so probably the consumer of `certstore` should configure it with a location to use. This is I think one of the uglier bits of nss. I don't know if it's possible to run NSS without a directory - obviously if your private key isn't stored on some kind of hardware token it's got to live somewhere...

btoews replied:
Gotcha. I had been thinking it would be normal for multiple applications to share this data, but I guess that isn't the case.

rhowe replied:
I think we can get away with being a "Mode 3, Type B" application as described here: https://wiki.mozilla.org/NSS_Shared_DB

Unfortunately, all the examples there are for other scenarios :face_with_head_bandage:

rhowe replied:
Also note that NSS databases can be password protected, so that will be fun to deal with

rhowe replied:
As far as I can tell, in general nobody attempts to have a system-wide NSS database and instead apps have a local one (e.g. my Debian unstable machine has nothing in /etc/ca-certificates/update.d for managing an NSS database, just Java and Mono). Whilst it's possible for apps to share an NSS database in the new SQLite world, I still don't see apps actually doing it (there's a [HOWTO guide](https://wiki.mozilla.org/NSS_Shared_DB_Howto) explaining how to set it up manually, involving making one in ~/.pki/nssdb and then symlinking it into the firefox & thunderbird profile directories). So, I think probably apps using certstore e.g. smimesign should take care of deciding where the database should live.

It's also not clear to me whether uses of hardware PKCS#11 tokens even need a database - the [FAQ](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/PKCS11/FAQ) suggests that everything happens in an in-memory database (see "Will NSS get the user's CA certificate via PKCS #11 and push it into the CA certificate database or is the CA certificate database expected to obtain the CA certificate by some other means?") so possibly we only need to use an on-disk DB if the soft ware PKCS#11 library (I think it's called softokn) is being used.

btoews replied:
Well, that's good news regarding PKCS#11 tokens at least. Applications having to maintain their own database isn't ideal, but isn't the end of the world either.

btoews reviewed (commented):
Almost there! This is coming together really nicely.

rhowe committed:
Improve behaviour if PK11_ListCerts fails

We don't need to shut everything down if we can't list the user's certificates
(in fact, this method only returns NULL if it fails to allocate memory for the
certificate list structure).

rhowe committed:
Improve error messages during Import()

rhowe commented:
> I'm wondering if we should make NSS an option on all platforms, since it's cross platform already. Maybe we could define an `NSSOpen()` method on all platforms and then have `openStore()` call `NSSOpen()` on Linux (make NSS the default on Linux).

I'm starting to wonder if we should use libressl or opensc directly instead on linux :joy:

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,324 @@
+		if C.PORT_GetError() == C.SEC_ERROR_INVALID_ALGORITHM {
+			return nil, ErrUnsupportedHash
+		}
+		return nil, fmt.Errorf("error signing: %s\n", C.GoString(C.GetErrorString()))
```
We don't need the trailing `\n`, since this is an error string (caught by go-lint).

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,324 @@
+}
+
+// Close implements the Store interface.
+func (ctx nssStore) Close() {
```
This should be consistent with the receiver name used above `func (store nssStore) ...` (caught by go-lint).

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,324 @@
+// openStore opens the current user's NSS database.
 func openStore() (Store, error) {
-	return nil, errors.New("certstore only works on macOS and Windows")
+	homeDir, err := os.UserHomeDir()
```
Using `os.Getenv("HOME")` instead of `os.UserHomeDir()` gets CI passing on older Go versions. `os.UserHomeDir()` was added in Go 1.12.0.

rhowe replied:
Yeah I was waiting to see how much backwards compatibility you wanted to have :)

UserHomeDir is more a cross-platform abstraction API so if we only care about Linux there's no need to use it

btoews replied:
:+1: For now, let's just do `os.Getenv("HOME")`.

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,324 @@
+	if p12 == nil {
+		return fmt.Errorf("error initialising PKCS#12 decoding: %s", C.GoString(C.GetErrorString()))
+	}
+	decoded := C.SEC_PKCS12DecoderUpdate(p12, (*C.uchar)(unsafe.Pointer(&data[0])), C.size_t(len(data)))
```
Using `C.ulong(len(data))` instead of `C.size_t(len(data))` gets CI passing on old Go versions. These types of changes are the most frustrating thing about working with Cgo 😡

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,324 @@
+	defer C.free(unsafe.Pointer(nssdbC))
+	defer C.free(unsafe.Pointer(emptyStringC))
+
+	ctx := C.NSS_InitContext(nssdbC, emptyStringC, emptyStringC, emptyStringC, nil, C.uint(0))
```
Another annoying Cgo difference: `C.PRUint32(0)` instead of `C.uint(0)` gets this building on older versions of Go.

btoews reviewed (commented):
A few nits from `go-lint` as well as suggestions for getting the TravisCI build working on Linux with older Go versions.

mastahyeti commented:
I was playing with this branch locally and got it to find the certificates on my Yubikey.

First, I had to tell NSS about the Yubico PKCS#11 library:

`modutil -dbdir ~/.pki/nssdb -add ykcs11 -libfile /usr/local/Cellar/yubico-piv-tool/1.6.2/lib/libykcs11.dylib -force`

And then I had to manually add the certificate to the database:

`certutil -A -i ~/Desktop/a.crt -d sql:~/.pki/nssdb -n "YubikeyAuthentication" -tu`

Certstore is now including the certificate in the returned identities, but I'm still getting an error trying to sign with it.

mastahyeti commented:
Also, this works on macOS. [Here's](https://github.com/mastahyeti/certstore/commit/f08783b8e5189d21d66305e7f78b5ca75c0d731c?w=1) the change I made to add a cross platform `OpenNSS` function. Test pass locally on macOS, but obviously not in CI, since NSS isn't installed there. Once this PR merges I can circle back to get that part working.

rhowe commented:
> Also, this works on macOS. [Here's](https://github.com/mastahyeti/certstore/commit/f08783b8e5189d21d66305e7f78b5ca75c0d731c?w=1) the change I made to add a cross platform `OpenNSS` function. Test pass locally on macOS, but obviously not in CI, since NSS isn't installed there. Once this PR merges I can circle back to get that part working.

I can see you're having fun :)

I'm about 5 beers deep this evening so I'll take a look another day!

mastahyeti commented:
> I'm about 5 beers deep this evening so I'll take a look another day!

Totally. Let's ship this for Linux before putting effort into NSS for other platforms.

rhowe committed:
Use ulong instead of size_t for compatibility with older go

rhowe commented:
OK, well we have a green travis build now, so that's a start. Once we get to merging this, I think I want to squash all the commits down to one with credit to the original author in the commit message - what do you think?

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,324 @@
+	SECItem *ret_nick = NULL;
+	CERTCertificate *cert = (CERTCertificate *)wincx;
+	if (!cancel || !cert) {
+		fprintf(stdout, "cancel: %p, cert: %p\n", cancel, cert);
```
Let's remove all the debugging output from this function.

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,324 @@
+		return NULL;
+	}
+	ret_nick->data = (unsigned char *)nick;
+	ret_nick->len = PORT_Strlen(nick);
```
Should this be `PORT_Strlen(nick) + 1`, since `nick` is null terminated?

rhowe replied:
Yep, it should. The test above then needs adjusting.

btoews reviewed (commented):
A few more small comments. I'm trying to find someone with more NSS experience to give it another once over.

mastahyeti commented:
> Once we get to merging this, I think I want to squash all the commits down to one with credit to the original author in the commit message - what do you think?

Yes! I'll defer to your judgement on how to rebase/squash things, but I agree that fewer commits would be good. Also creditting the original author would be ideal. If you want to squash it down to a single commit, you could list yourself as a co-author ([Co-Authored-By docs](https://help.github.com/en/articles/creating-a-commit-with-multiple-authors)).

rhowe committed:
Clean up nickname collision callback

philandstuff commented:
I've built smimesign locally with this branch of certstore and I can't get it to actually sign.

I've added the OpenSC module to my NSS database, and my yubikey PIV cert appears when I run `smimesign --list-keys`, but if I try to do `smimesign -s` then the call to `C.PK11_FindKeyByAnyCert(&cert, nil)` returns `nil` for some reason.

I'm still trying to understand why but I wanted to ask: have either of you managed to get this working with smimesign?

philandstuff commented on certstore_linux.go:
```diff
@@ -1,14 +1,320 @@
+	}
+	var (
+		cert = C.CERTCertificate(*i)
+		key  = C.PK11_FindKeyByAnyCert(&cert, nil)
```
The docs for `PK11_FindKeyByAnyCert` say:

>  When you are finished with the private key structure returned by PK11_FindKeyByAnyCert, you must free it by calling SECKEY_DestroyPrivateKey.

but I don't see a call to SECKEY_DestroyPrivateKey here?  It'd probably be best to do it in a `defer` immediately after we've confirmed we didn't get a `nil` response.

rhowe replied:
Good catch. Fixed.

philandstuff commented:
ah I'm wondering if I'm hitting the problem mentioned above

> Signing with the 9e slot works! This is the only slot that doesn't require a pin. The slots that do require a pin still don't work for signing.

rhowe committed:
Ensure we free private keys allocated by PK11_FindKeyByAnyCert

rhowe commented:
You might find you can log in to the card with something like `pkcs11-tool` from opensc and get a PIN entered before then using another slot..

philandstuff commented:
Either that or pass a password function callback to `PK11_FindKeyByAnyCert`?

philandstuff commented:
But in any case that can surely wait for a further PR?

rhowe commented:
> Either that or pass a password function callback to `PK11_FindKeyByAnyCert`?

That was the kind of appropriate thing I was thinking of, yes.. something that ultimately runs a pinentry program or similar I guess, and yes IMO that can be a later addition - I still haven't actually done as much as you have, building smimesign against this branch and testing it on real hardware.

philandstuff commented:
If you already have certstore checked out, you can use a go.mod `replace` directive in smimesign to use your local certstore checkout:

https://github.com/philandstuff/smimesign/commit/3a66dc52cf0c6c0cda8337eece4dc51e3f0cb2e6#diff-37aff102a57d3d7b797f152915a6dc16R17

Then building smimesign with the local certstore is just `go build` - couldn't be easier!

rhowe commented:
> Either that or pass a password function callback to `PK11_FindKeyByAnyCert`?

Actually, reading the docs more closely that's not what that argument is for.
My understanding is that you call [PK11_SetPasswordFunc](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/PKCS11_Functions#PK11_SetPasswordFunc) to register a callback function for collecting the PIN needed to unlock the card. This function returns the PIN as a string, but its arguments are a pointer to a slot structure referencing the card, a boolean indicating whether this is a first attempt at a login or not and a pointer to some data which means something to your function.

The pointer passed to `PK11_FindKeyByAnyCert` is this latter entity - it's up to your password function to know what to do with it - maybe for a GUI app it could hold a reference to the active window so that any dialog boxes are properly parented, maybe it's a location of a socket to make a request to etc.

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,321 @@
+	}
+	if (old_nick && old_nick->data && old_nick->len &&
+		PORT_Strlen(nick) == old_nick->len - 1 &&
+		!PORT_Strncmp((char *)old_nick->data, nick, old_nick->len - 1)) {
```
I'm just realizing that this whole function was copied from NSS ([source](https://github.com/nss-dev/nss/blob/master/cmd/pk12util/pk12util.c#L123-L170)). There, they seem to assume that `old_nick->len` doesn't account for a trailing null byte. Have you managed to test this? I'm wondering if this is a bug in NSS or if we're misunderstanding how strings in `SECItem`s work.

btoews replied:
I was a bit unsure about the legality of copy-pasting from the NSS repo. I checked with GitHub's legal team and they say this is dangerous. This function should only come into play when trying to import something with a non-unique nickname, right? It seems like we could get by without the logic in this function for now. Could we change this function to just `return NULL` for now?

rhowe replied:
I wasn't really sure how much this function was needed either.. I'll have a ponder

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,321 @@
+
+// Close implements the Identity interface.
+func (i *nssIdentity) Close() {
+	cert := C.CERTCertificate(*i)
```
In a few places, you're doing `C.CERTCertificate(*i)`, and then doing something with `&cert`. Am I correct in thinking that this is copying the struct and then working on the copy? If so, it seems that something like this `CERT_DestroyCertificate` call would only affect the reference count on the copy. It seems like this could be problematic if we were to then accidentally try using the original struct again?

I'm wondering if, instead, it would be possible to convert the `*nssIdentity` directly to a `*C.CERTCertificate`. Does something like `cert := (*C.CERTCertificate)(i)` work?

rhowe replied:
I think you're right and indeed that does work. Amendments coming right up..

btoews commented on certstore_linux.go:
```diff
@@ -1,14 +1,321 @@
+	}
+	defer C.CERT_DestroyCertList(certs)
+	for node = C.CertListHead(certs); C.CertListEnd(node, certs) == 0; node = C.CertListNext(node) {
+		identity := nssIdentity(*C.CERT_DupCertificate(node.cert))
```
Similar problem here, I think. Could this be `(*nssIdentity)(C.CERT_DupCertificate(node.cert))` to avoid copying the struct?

rhowe replied:
Maybe. I think I tried something like that and got shouted at by cgo but I'll have a play.

btoews reviewed (commented):
A few more comments. Let me know if you're getting tired of working on this PR. I'd be happy to try making some of the suggested changes myself and pushing them to this branch.

rhowe committed:
Remove redundant variable declaration

ListsOfArrays commented:
Hello!
There are a couple issues identified with this branch, one fixed would it be ready for merge?

mateusz-klatt commented:
Hi,

I'm the autor of #11 - sorry was out of the gird for a while... what's the status of this PR?

rhowe commented:
> Hi,
> 
> I'm the autor of #11 - sorry was out of the gird for a while... what's the status of this PR?

It's kinda stagnated, as you can see. If I remember rightly, one of the main remaining questions is how to manage and initialise the NSS db, otherwise I _think_ the code is mostly working. I tried to preserve your initial MR and then add isolated cleanup commits on top.

If you know NSS well, that would be great input to have - I make no pretence of being a go or an nss expert.

I haven't actually used this to successfully sign a git commit though, so there may still be work to do.

Maybe take a look at the commits we've piled on top of your initial work and see if they look reasonable? I'm happy to spend some time revisiting this - I'll try and rebase against `main` soon.

mateusz-klatt commented:
Hi,

Actually I was able to sign commits with embedded git extensions - so it’s more multi platform approach - however not as easy as smimesign - still prefer certs over gpg as it is more corporate friendly way of commits verification - at this stage I’m not going to engage into open source projects for a while, but as soon as I will be able I’m happy to look into it again. 

Kind Regards,

Mateusz Klatt

> Wiadomość napisana przez Russell Howe <notifications@github.com> w dniu 12.01.2021, o godz. 11:32:
> 
> ﻿
> Hi,
> 
> I'm the autor of #11 - sorry was out of the gird for a while... what's the status of this PR?
> 
> It's kinda stagnated, as you can see. If I remember rightly, one of the main remaining questions is how to manage and initialise the NSS db, otherwise I think the code is mostly working. I tried to preserve your initial MR and then add isolated cleanup commits on top.
> 
> If you know NSS well, that would be great input to have - I make no pretence of being a go or an nss expert.
> 
> I haven't actually used this to successfully sign a git commit though, so there may still be work to do.
> 
> Maybe take a look at the commits we've piled on top of your initial work and see if they look reasonable? I'm happy to spend some time revisiting this - I'll try and rebase against main soon.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

lgarron commented:
This project is now part of `smimesign`.
Please send this PR to https://github.com/github/smimesign if it's still relevant!
